src/include/miopen/optim/adam/problem_description.hpp

```diff
@@ -0,0 +1,181 @@
+           (expAvgInDesc->GetType() != dtype) || (expAvgOutDesc->GetType() != dtype) ||
+           (expAvgSqInDesc->GetType() != dtype) || (expAvgSqOutDesc->GetType() != dtype) ||
+           (maxExpAvgSqInDesc != nullptr && maxExpAvgSqInDesc->GetType() != dtype) ||
+           (maxExpAvgSqOutDesc != nullptr && maxExpAvgSqOutDesc->GetType() != dtype))
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "Adam: Tensor types do not match.");
+        }
+
+        if(is_amp && (paramOutFloat16Desc != nullptr) &&
+           (paramOutFloat16Desc->GetType() != miopenHalf))
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "Adam: Invalid type of param_out_float16.");
+        }
+
+        auto numel = paramInDesc->GetElementSize();
+        if((paramOutDesc->GetElementSize() != numel) || (gradInDesc->GetElementSize() != numel) ||
+           (expAvgInDesc->GetElementSize() != numel) ||
+           (expAvgOutDesc->GetElementSize() != numel) ||
+           (expAvgSqInDesc->GetElementSize() != numel) ||
+           (expAvgSqOutDesc->GetElementSize() != numel) ||
+           (is_amp && paramOutFloat16Desc != nullptr &&
+            paramOutFloat16Desc->GetElementSize() != numel) ||
+           (maxExpAvgSqInDesc != nullptr && maxExpAvgSqInDesc->GetElementSize() != numel) ||
+           (maxExpAvgSqOutDesc != nullptr && maxExpAvgSqOutDesc->GetElementSize() != numel))
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "Adam: Tensor dimension lengths do not match.");
+        }
+    }
+
+    const TensorDescriptor& GetParamDesc() const { return *paramInDesc; }
+    const TensorDescriptor& GetGradDesc() const { return *gradInDesc; }
+    bool IsAmp() const { return is_amp; }
+    bool IsAdamW() const { return adamw; }
+    bool IsStepHost() const { return (stepInDesc == nullptr); }
+    bool IsAllPacked() const
+    {
+        if(!(paramInDesc->IsPacked() && gradInDesc->IsPacked() && expAvgInDesc->IsPacked() &&
+             expAvgSqInDesc->IsPacked()))
+            return false;
+        return true;
+    }
+
+    NetworkConfig MakeNetworkConfig() const override;
+
+private:
+    const TensorDescriptor* paramInDesc         = nullptr;
+    const TensorDescriptor* paramOutDesc        = nullptr;
+    const TensorDescriptor* gradInDesc          = nullptr;
+    const TensorDescriptor* expAvgInDesc        = nullptr;
+    const TensorDescriptor* expAvgOutDesc       = nullptr;
+    const TensorDescriptor* expAvgSqInDesc      = nullptr;
+    const TensorDescriptor* expAvgSqOutDesc     = nullptr;
+    const TensorDescriptor* paramOutFloat16Desc = nullptr;
+    const TensorDescriptor* maxExpAvgSqInDesc   = nullptr;
+    const TensorDescriptor* maxExpAvgSqOutDesc  = nullptr;
+    const TensorDescriptor* gradScaleDesc       = nullptr;
+    const TensorDescriptor* foundInfDesc        = nullptr;
+    const TensorDescriptor* stepInDesc          = nullptr;
+    const TensorDescriptor* stepOutDesc         = nullptr;
```

Comment:
IIRC normally we store copies of tensor descriptors in the problem description. These are lightweight objects

Reply:
[resolved]